Source/JavaScriptCore/runtime/WaiterListManager.h

```diff
@@ -0,0 +1,139 @@
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include "JSGlobalObject.h"
+#include "JSPromise.h"
+#include "runtime/DeferredWorkTimer.h"
+#include <wtf/DataLog.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/ParkingLot.h>
+#include <wtf/RawPointer.h>
+
+namespace JSC {
+
+class WaiterListManager {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    static WaiterListManager& singleton();
+
+    void addWaiter(void* ptr, JSPromise* promise, Seconds timeout)
+    {
+        RELEASE_ASSERT(ptr && promise);
+        VM& vm = promise->globalObject()->vm();
+        vm.apiLock().lock();
+        auto ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
+        vm.apiLock().unlock();
+
+        ParkingLot::parkAsync(ptr, ticket);
+
+        m_registeredVMs.add(&vm);
+        RunLoop::current().dispatchAfter(timeout, [ptr, ticket] {
+            WaiterListManager::singleton().timeout(ptr, ticket);
+        });
+    }
+
+    void timeout(void* ptr, DeferredWorkTimer::Ticket ticket)
+    {
+        RELEASE_ASSERT(ptr && ticket);
+        JSPromise* promise = jsCast<JSPromise*>(ticket->target());
+        VM* vm = &promise->globalObject()->vm();
+        if (!m_registeredVMs.contains(vm))
+            return;
+
+        ParkingLot::unparkAsync(ptr, ticket, [this](void* data) {
+            bool isOKString = false;
+            notifyWaiterImpl(data, isOKString);
+        });
+    }
+
+    unsigned notifyWaiter(void* ptr, unsigned count)
+    {
+        RELEASE_ASSERT(ptr);
+        bool isAsyncMode = true;
+        unsigned notified = ParkingLot::unparkCount(
+            ptr, count, [this](void* data) {
+                bool isOKString = true;
+                notifyWaiterImpl(data, isOKString);
+            },
+            isAsyncMode);
+        return notified;
+    }
+
+    void notifyWaiterImpl(void* data, const bool isOKString)
```

Comment:
nit: If I was authoring this I would have gone with `enum DeferredWorkResult : bool { OK, Timeout }` or something.